x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/normalizer/ShortCircuitingRenormalizer.java

```diff
@@ -122,7 +122,26 @@ private boolean tryFinishWork() {
     }
 
     private void forceFinishWork() {
-        semaphore.release();
+        // We cannot allow new quantiles to be added while we are failing from a previous renormalization failure.
+        synchronized (quantilesDeque) {
```

Comment:
To prevent the scenario in the description, we need to `countDown` ALL the latches stored in the `quantilesDeque`.

Keeping the earliest quantile, which already has a tripped latch does mean the following:

- `latchDeque` now has a latch in it that is already countedDown (previously possible)
- `quantilesDeque` now has a latch in it that is already counted down (not previously possible)

I think `quantilesDeque` having an already counted down latch is OK in this scenario. `doRenormalizations` will not get called again until `renormalize` calls it. This would imply that a new quantile with a new latch is seen. The worst thing here is that the previously counted down latch is counted down again when `getLatestQuantilesWithLatchAndClear` is called in `doRenormalizations` which is a no-op.